R/chat.R

```diff
@@ -506,33 +547,68 @@ Chat <- R6::R6Class(
       user_turn,
       stream,
       echo,
-      tool_mode
+      tool_mode = "concurrent",
+      yield_as_content = FALSE
     ) {
       tool_errors <- list()
       withr::defer(warn_tool_errors(tool_errors))
 
       while (!is.null(user_turn)) {
-        # fmt: skip
-        for (chunk in await_each(private$submit_turns_async(user_turn, stream = stream, echo = echo))) {
+        assistant_turn <- private$submit_turns_async(
+          user_turn,
+          stream = stream,
+          echo = echo
+        )
+        for (chunk in await_each(assistant_turn)) {
+          if (yield_as_content && is.character(chunk)) {
+            chunk <- ContentText(chunk)
+          }
           yield(chunk)
         }
 
-        tool_calls <- invoke_tools_async(
-          self$last_turn(),
-          echo = echo,
-          on_tool_request = private$callback_on_tool_request$invoke_async,
-          on_tool_result = private$callback_on_tool_result$invoke_async
-        )
-        if (tool_mode == "sequential") {
-          tool_results <- list()
-          for (result in coro::await_each(tool_calls)) {
-            tool_results <- c(tool_results, list(result))
+        user_turn <- NULL
+        if (turn_has_tool_request(self$last_turn())) {
```

Comment:
Can we use `assistant_turn` instead of `self$last_turn()`? It feels a bit weird to reach back up here.

Reply:
Actually, no, `assistant_turn` here is the generator from `private$submit_turns()` and `self$last_turn()` is the full `Turn` object.

I used `assistant_turn <- private$submit_turns_async()` to avoid overflowing the line inside the `for` loop, but I should probably find a better name. Do you have a naming convention or preference for generators?

Reply:
In f58a07e9ed72335316a8e7e9288ee75590637b8a, I updated the naming to `assistant_chunks` and later used `assistant_turn <- self$last_turn()`.